modules/kafka/tests/test_kafka_topic.c

```diff
@@ -0,0 +1,182 @@
+  cr_assert_eq(kafka_dd_validate_topic_name(driver, param->topic_name), param->isvalid);
+  log_pipe_unref((LogPipe *)driver);
+}
+
+Test(kafka_topic, test_resolve_template_topic_name)
+{
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
+
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "$MSG", NULL);
+  kafka_dd_set_topic(driver, topic_name);
+  kafka_dd_set_fallback_topic(driver, "fallbacktopicname");
+
+  g_assert(kafka_dd_init((LogPipe *) driver));
+
+  KafkaDestDriver *kafka_driver = (KafkaDestDriver *) driver;
+
+  KafkaDestWorker *worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
+
+  LogMessage *msg = log_msg_new_empty();
+  log_msg_set_value_by_name(msg, "MSG", "valid_template_topic", -1);
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "valid_template_topic");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
+
+  log_msg_set_value_by_name(msg, "MSG", "invalid name", -1);
+
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "fallbacktopicname");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
+  log_msg_unref(msg);
+
+  log_template_unref(topic_name);
+  log_threaded_dest_worker_free(&worker->super);
+  log_pipe_unref((LogPipe *)driver);
+  cfg_free(configuration);
+}
+
+Test(kafka_topic, test_calculate_topic_from_template)
+{
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
+
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "$MSG", NULL);
+  kafka_dd_set_topic(driver, topic_name);
+  kafka_dd_set_fallback_topic(driver, "fallbackhere");
+
+  g_assert(kafka_dd_init((LogPipe *) driver));
+
+  KafkaDestDriver *kafka_driver = (KafkaDestDriver *) driver;
+
+  KafkaDestWorker *worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
+
+  LogMessage *msg = log_msg_new_empty();
+  log_msg_set_value_by_name(msg, "MSG", "validtopic", -1);
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)), "validtopic");
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)),
+                   rd_kafka_topic_name(_calculate_topic(worker, msg)));
+
+  log_msg_set_value_by_name(msg, "MSG", "invalid name", -1);
+
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)), "fallbackhere");
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)),
+                   rd_kafka_topic_name(_calculate_topic(worker, msg)));
+  log_msg_unref(msg);
+
+  log_template_unref(topic_name);
+  log_threaded_dest_worker_free(&worker->super);
+  log_threaded_dest_driver_free((LogPipe *)driver);
```

Comment:
the log_threaded_dest_driver_free() is not needed here and is taken care by the subsequent log_pipe_unref()

Reply:
the worker needs to be freed though as it is not managed by the KafkaDestDriver instance.